### PR TITLE
[WIP] design: sketch some notes on txn protocol

### PR DESCRIPTION
**Pull request by sploiselle**
Collecting context/description of some conversation we've been having around our transaction protocol

### CODE REVIEW

File: doc/developer/design/aphyr-txn-notes.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+# CAS
+
+## Background
+
+This is the design Kyle has been moving forward with fleshing out, which relies
+on a the WAL having a timestamp only ever written to atomically.
+
+## Materialize design
+
+Each writable TVC will have a WAL that acts as a buffer for writes before being
```
Comment (mjibson) [inline doc/developer/design/aphyr-txn-notes.md:76]: Oop! I think it's that there is still a single WAL, not one per TVC. The WAL appends all writes an a transaction atomically, then storage nodes are responsible for consuming it. The difference between the CAS and non CAS versions is that in the CAS version, if a txn exists in the WAL it committed. In the non CAS version consumers of the WAL must do something to decide if the txn committed (I think this was to have in the txn data a "previous WAL operation timestamp must be X" value which can checked in the WAL, or maybe some additional coordination between all storage nodes (I'm less sure about this)).

File: doc/developer/design/aphyr-txn-notes.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+committed to storage. Each WAL will have a timestamp indicating the last time at
+which an operation occurred; we'l call this the WAL ts for simplicity.
+
+When users want to write to a TVC, they'll use `append(updates, from ts, new
+frontier)`. This implementation of `append` will succeed iff it's able to update
+the WAL's ts using an atomic CAS, i.e. the WAL's current ts is `from ts`.
+
+When reading, the ADAPTER consults the WAL and gets its current timestamp
```
Comment (mjibson) [inline doc/developer/design/aphyr-txn-notes.md:84]: I do not recall discussing this, that reads ever need to consult the WAL for a timestamp.

File: doc/developer/design/aphyr-txn-notes.md (hunk 1 above)
Comment (sploiselle) [reply doc/developer/design/aphyr-txn-notes.md:76]: This was totally right and an oversight on my part.

**Comment by sploiselle** [comment]
Matt is going to take a crack at distilling this for the SQL standup next week.

**Comment by sploiselle** [comment]
Further conversation in Slack has also yielded takes that obviate the approaches discussed here.

**Pull request closed without merging**